cloud-controller-manager/do/loadbalancers.go

```diff
@@ -275,6 +274,56 @@ func (l *loadBalancers) EnsureLoadBalancer(ctx context.Context, clusterName stri
 	}, nil
 }
 
+func (l *loadBalancers) updateLoadBalancer(ctx context.Context, lb *godo.LoadBalancer, lbRequest *godo.LoadBalancerRequest, service *v1.Service) (*godo.LoadBalancer, error) {
+	var lbCertID string
+	for _, rule := range lb.ForwardingRules {
+		if rule.TlsPassthrough {
+			continue
+		}
+		if rule.CertificateID != "" {
+			lbCertID = rule.CertificateID
+			// it's not clear that it is even possible for different forwarding rules
+			// on the same loadbalancer to have different certificates associated
+			// with them -- as far as we are concerned where a load balancer is
+			// created by CCM in the first place it doesn't seem possible so we only
+			// need to grab the first one we find
+			break
+		}
+	}
+
+	checkServiceCertID := true
+	if lbCertID != "" && lbCertID != getCertificateID(service) {
+		lbCert, _, err := l.resources.gclient.Certificates.Get(ctx, lbCertID)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get current certificate for lb")
+		}
+
+		if lbCert.Type == certTypeLetsEncrypt {
+			l.ensureCertificateIDAnnot(service, lbCertID)
+			checkServiceCertID = false
+		}
+	}
+
+	if checkServiceCertID {
+		_, _, err := l.resources.gclient.Certificates.Get(ctx, getCertificateID(service))
+		if err != nil {
+			respErr, ok := err.(*godo.ErrorResponse)
+			if ok && respErr.Response.StatusCode == http.StatusNotFound {
+				fmtStr := "certificate UUID %q not found; the %q service annotation refers to nonexistent DO Certificate"
+				respErr.Message = fmt.Sprintf(fmtStr, getCertificateID(service), annDOCertificateID)
```

Comment:
I think we shouldn't overwrite the error message from an error we do not own. Creating a new error seems like the better option to me.

Doing so would also allow us to replace the type assertion by checking the response directly like

```
if resp != nil && resp.StatusCode == http.StatusNotFound {
  return fmt.Errorf(...)
}
```

(This would replace line 310 forward.)